### PR TITLE
Fix migration - corrupt org namespace

### LINKED ISSUE #792: [ver. 0.14.59] Duplicate naming in output package
As of `apibuilder ver 0.14.59`, I noticed that the package naming has changed, and a duplication got somehow introduced. I presume that it wasn't intentional.

Affected generators: Play, Http4s, Scala Model, Java Models and perhaps more.

Example:
Project: https://app.apibuilder.io/michal/state-example/latest (public sample project)

Project name: state-example
File: https://app.apibuilder.io/michal/state-example/0.0.1-dev/http4s_0_20/MichalStateExampleV0ModelsJson.scala

*Actual* package line (with header comment):
```
/**
 * Generated by API Builder - https://www.apibuilder.io
 * Service version: 0.0.1-dev
 * apibuilder 0.14.59 app.apibuilder.io/michal/state-example/0.0.1-dev/http4s_0_20
 */
package io.github.mkows.state.example.v0.state.example.v0.models {
```

*Expected* package line:
```
package io.github.mkows.state.example.v0.models {
```

The same for *Play* https://app.apibuilder.io/michal/state-example/0.0.1-dev/play_2_x_json/MichalStateExampleV0Models.scala which starts with:
```
/**
 * Generated by API Builder - https://www.apibuilder.io
 * Service version: 0.0.1-dev
 * apibuilder 0.14.59 app.apibuilder.io/michal/state-example/0.0.1-dev/play_2_x_json
 */
package io.github.mkows.state.example.v0.state.example.v0.models {
```

It seem to affect many more projects.

I can confirm that expected behavior was there 9-10 days ago for sure, and perhaps even couple days ago as well. I bet that the issue must have been caused by some very recent deployment (strangely enough, files generated few days ago -- that have the correct package -- have the same apibuilder version i.e. `0.14.59`).

CC: @gheine

### PR DESCRIPTION
**Pull request by mkows**
Fixes #792.

Currently migration overwrites with `org namespace` with `service namespace`.
Hence  
`org namespace=io.github.mkows` and application `playground-api`, the result is not `io.github.mkows.playground.api.v0` (as expected) but `io.github.mkows.playground.api.v0.playground.api.v0`.

This PR introduces `1 + N` db calls (up from `1`) to fetch organization namespace -- this is to not modify `Version` model (an alternative would be to have a dedicated `Version + org namespace` model).

h3. Initial description (archive)

Unit test demonstrates the issue.
Proposal of the fix -- not tested at all yet -- just code draft.

```
records.foreach { version =>
        val orgKey = version.organization.key
        val applicationKey = version.application.key
        val versionName = version.version
        val versionGuid = version.guid
        // ...
        Logger.error(s"DEV: Migrating for version.service.namespace=${version.service.namespace}, orgKey=${version.organization.key}, orgGuid=${version.organization.guid}")

        Logger.info(s"Migrating $orgKey/$applicationKey/$versionName versionGuid[$versionGuid] to latest API Builder spec version[$ServiceVersionNumber]")

        // current PROD code:
        val config = ServiceConfiguration(
          orgKey = orgKey,
          orgNamespace = version.service.namespace,
          version = versionName
        )

```

prints
```
DEV: Migrating for version.service.namespace=io.github.mkows.playground.api.v0, orgKey=michal, orgGuid=ceeea5e9-d2f8-4be2-967e-5e9a2fe6b260
``` 
which demonstrates how currently on PROD `orgNamespace` gets overwritten with _full_ `version.service.namespace` while it should be unaltered (i.e. `io.github.mkows`)

### CODE REVIEW

**Comment by mkows** [comment]
Update:
alright, managed to reproduce the issue locally. For `master` branch, getting:
```
/**
 * Generated by API Builder - https://www.apibuilder.io
 * Service version: 0.0.1-dev
 * apibuilder 0.14.75 localhost 9000/michal/playground-api/0.0.1-dev/play_2_6_client
 */
package io.github.mkows.playground.api.v0.playground.api.v0.models {
```
With the proposed fix, it's:
```
/**
 * Generated by API Builder - https://www.apibuilder.io
 * Service version: 0.0.1-dev
 * apibuilder 0.14.75 localhost 9000/michal/playground-api/0.0.1-dev/play_2_6_client
 */
package io.github.mkows.playground.api.v0.models {
``` 🎉

**Comment by mkows** [comment]
> i think we need to use the service namespace as in the implementation we allow users to specify their own namespace in the service spec (org namespace is used as the default)

This one is already handled here:
https://github.com/apicollective/apibuilder/blob/master/core/src/main/scala/core/builder/api_json/ServiceBuilder.scala#L31

`internal.namespace` retains the namespace specified by the user.
It's the one coming from `select * from cache.services` (passed there via `version.service`).


So for the current version (as on PROD) example values are:
**For an app with namespace prespecified**

```
config=ServiceConfiguration(
michal,io.github.mkows.playground-hardcoded-namespace,0.0.1-dev), 
version.service.namespace=io.github.mkows.playground-hardcoded-namespace
```

and then inside `ServiceBuilder`:
```
internal.namespace=Some(io.github.mkows.playground-hardcoded-namespace), and 
config.applicationNamespace(key)=io.github.mkows.playground-hardcoded-namespace.playground.apj.v0
```
(not that here again `applicationNamespace` is -- while being used -- corrupt)

After this PR the above (modulo slight change in app name) changes to:
```
config=ServiceConfiguration(michal,io.github.mkows,0.0.1-dev))
```
and inside `ServiceBuilder`:
```
internal.namespace=Some(io.github.mkows.playground-apk-hardcoded-namespace), and 
config.applicationNamespace(key)=io.github.mkows.playground.apk.v0
```
**NOTE** that both values are correct -- and still only `internal.namespace` is used.

----

Back to the issue... This issue only affects apps with no namespace pre-specified, and is caused by the fact that `version.service.namespace` is assigned to `orgNamespace`
as a result, when `config.applicationNamespace(key)` is called, the result is corrupt (as org name already contains full `version.service.namespace`) -- and yeah, that value comes from `cache.services` / internal service model and therefore it's not empty (despite being absent in api.json).

This PR fixes the issue.

As a bonus, it will auto-heal all currently broken services on PROD -- as application namespace (and consequently `service > namespace`) will get re-calculated to the correct value.

----
btw. in order to lookup broken services on PROD run `select * from cache.services` and grep for `namespace` -- this is where the corrupt values live.

**Comment by mkows** [comment]
(I pushed my dev code to `792-master-dev-notes` branch -- to share my code used to troubleshoot on master branch. in order to enforce migration, just bump client version `generated/app/ApicollectiveApibuilderSpecV0Client.scala` around `line 1533`).

This PR is ready for the review.

File: api/app/db/VersionsDao.scala [hunk 1]
```diff
@@ -356,17 +356,24 @@ class VersionsDao @Inject() (
         val versionName = version.version
         val versionGuid = version.guid
 
-        Logger.info(s"Migrating $orgKey/$applicationKey/$versionName versionGuid[versionGuid] to latest API Builder spec version[$ServiceVersionNumber]")
+        val organizationNamespace = {
+          SQL("select namespace from organizations where guid = {orgGuid}::uuid")
+            .on("orgGuid" -> version.organization.guid)
+            .executeQuery()
+            .as(SqlParser.scalar[String].single)
+        }
```
Comment (gheine) [inline api/app/db/VersionsDao.scala]: What does `. single` return if there is no record? Might be better to use `OrganizationsDao.findByGuid` and deal with the fact that the org may not exist.

**Comment by mkows** [comment]
Thanks @gheine. Very good point, I went low level as was deep down investigating psql while writing that code. Also handled org missing as failed migration (the only reason for that really would be that org was deleted after records were read -- but, still, it should not break the process but simply skip the current app).

## `tldr:`
The issue here is caused by org namespace being assigned service namespace value https://github.com/apicollective/apibuilder/pull/793/files#diff-d6793cef65a5b402b5bbf2157d518806L363 
Then:
1. for a service with pre-specified namespace, that's irrelevant as config.applicationNamespace is not used -- as `internal.namespace` takes precedence https://github.com/apicollective/apibuilder/blob/02547890a9ab7586736481211ae4a8da3864de49/core/src/main/scala/core/builder/api_json/ServiceBuilder.scala#L31
2. for a service with no pre-specified namespace, that's causing this bug as 
resulting `config.applicationNamespace` (effective) is evaluated to `<version.service.namespace>.<app-key-based-segment>` https://github.com/apicollective/apibuilder/blob/02547890a9ab7586736481211ae4a8da3864de49/lib/src/main/scala/ServiceConfiguration.scala#L17 (note: `version.service.namespace` above is calculated and provided even if not specified in `api.json`)
(CC: @mbryzek ^^^)

File: api/app/db/VersionsDao.scala (hunk 1 above)
Comment (mkows) [reply api/app/db/VersionsDao.scala]: Thanks! Addressed.

**Comment by sullis** [comment]
LGTM @mkows

**Comment by mbryzek** [comment]
@mkows pls ping if you need somebody to merge (thanks for the work here!)

**Pull request merged**